File: PCA/PCA.m
function [Y]=PCA(X,k)
    [m,n] = size(X);
    u = 0;
    for i = 1:m
        for j = 1:n
            u = u + X(i,j);
        end
        u = u/n;
        for j = 1:n
            X(i,j) = X(i,j) - u;
        end
        u = 0;
    end
    [eVe,eVa] = eig(X*X');
    eVa = diag(eVa);
    [~,index] = sort(eVa,'descend');
    P = eVe(:,index);
    P = P(:,1:k);
    Y = P'*X;
end

% clc;
% clear;
% 
% A = [1,2,3,4,5;6,7,8,9,10;11,12,13,14,14];
% [m,n] = size(A);
% u = 0;
% k=2;
% 
% for i = 1:m
%     for j = 1:n
%         u = u + A(i,j);
%     end
%     u = u/n;
%     for j = 1:n
%         A(i,j) = A(i,j) - u;
%     end
%     u = 0;
% end
% 
% [eVe,eVa] = eig(A*A');
% eVa = diag(eVa);
% [eVa,index] = sort(eVa,'descend');
% P = eVe(:,index);
% P = P(:,1:k);
% Y = P'*A;

File: PCA/PCA.py
import numpy as np
import math


def PCA(New_X, k):
    # New_X = np.copy(X)
    m, n = New_X.shape
    u = 0
    for i in range(m):
        for item in New_X[i, :]:
            u += item
        u /= n
        New_X[i, :] -= u
        u = 0.0
    S2, V = np.linalg.eig(np.dot(New_X, New_X.T))
    index_value_S2 = sorted(enumerate(S2), key=lambda x: x[1], reverse=True)
    V_index = []
    for index, value in index_value_S2:
        V_index.append(index)
    V = V[:, V_index]
    P = V[:, 0:k]
    Y = P.T @ New_X
    return Y, P


A = np.array([[1, 0, 2, 5, 3], [2, 0, 4, 7, 8], [3, 2, 7, 8, 9]], dtype=np.float64)
# A = np.array([[1, 0, 2], [5, 3, 2], [0, 4, 7], [8, 3, 2], [7, 8, 9]], dtype=np.float64)
Y, P = PCA(A, 2)
print(Y)
print(P)
